PRODDOCS-528: dbt.invoke (#8910)
fishtown-analytics/docs.getdbt.com: website/docs/reference/programmatic-invocations.md

title: "Programmatic invocations"
---

> In v1.5, <Constant name="core" /> added support for programmatic invocations. You can run the same top-level commands as the <Constant name="core" /> CLI from a Python script or application through the exposed entry point.

Reviewer: I don't think we need to call out the v1.5 support anymore since it's long deprecated. This whole first paragraph can probably be removed and combined with the second
```suggestion
```

> Programmatic invocations let you call dbt commands from Python, instead of running them in a shell. This is useful when you want to embed dbt runs into a larger application or workflow, while still using the same command surface area as the CLI.

Reviewer: ```suggestion
Programmatic invocations let you call dbt commands from Python scripts and applications, instead of running them in a shell. This is useful when you want to embed dbt runs into a larger application or workflow, while still using the same command surface area as the CLI.
```

    print(f"{r.node.name}: {r.status}")
```

> For implementation details, refer to the source definitions of `dbtRunner` and `dbtRunnerResult` in the [<Constant name="core" /> repository](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/cli/main.py).

Author: ```suggestion
For implementation details, refer to the source definitions of `dbtRunner` and `dbtRunnerResult` in the [<Constant name="core" /> repository](https://github.com/dbt-labs/dbt-core/blob/1.latest/core/dbt/cli/main.py).
```

- `exception`: When the dbt invocation encounters an unhandled error and does not complete, the exception that was raised.

There is a one-to-one correspondence between [CLI exit codes](/reference/exit-codes) and the `dbtRunnerResult` returned by a programmatic invocation:

| Scenario                                                                                    | CLI Exit Code | `success` | `result`         | `exception` |
|---------------------------------------------------------------------------------------------|--------------:|-----------|-------------------|-------------|
| Invocation completed without error                                                          | 0             | `True`      | varies by command | `None`        |
| Invocation completed with at least one handled error (for example, test failure or model build error) | 1             | `False`     | varies by command | `None`        |
| Unhandled error. Invocation did not complete, and returns no results.                       | 2             | `False`     | `None`              | Exception   |

## Commitments and caveats

> We're making an ongoing commitment to providing a Python entry point at functional parity with <Constant name="core" />'s CLI. We reserve the right to change the underlying implementation used to achieve that goal. We expect that the current implementation will unlock real use cases in the short- and medium-term while we work on a set of stable, long-term interfaces that will ultimately replace it.

Author: ```suggestion
We're making an ongoing commitment to providing a Python entry point at functional parity with <Constant name="core" />'s CLI. We reserve the right to change the underlying implementation used to achieve that goal. We expect that the current implementation will unlock real use cases in the short and medium-term while we work on a set of stable, long-term interfaces that will ultimately replace it.
```

    print(f"{r.node.name}: {r.status}")
```

For implementation details, refer to the source definitions of `dbtRunner` and `dbtRunnerResult` in the [<Constant name="core" /> repository](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/cli/main.py).

## Supported arguments

> `dbtRunner.invoke` accepts the same arguments as the <Constant name="platform_cli" />. The first positional argument is the subcommand (for example, `run`, `build`, `test`), followed by any flags and options you would normally pass on the command line.

Reviewer: ```suggestion
`dbtRunner.invoke` accepts the same arguments as the <Constant name="platform_cli" />. The first positional argument is the command (for example, `run`, `build`, `test`), followed by any flags and options you would normally pass on the command line.
```

> For example, `dbt.invoke(["run", "--select", "tag:my_tag"])` is equivalent to running `dbt run --select tag:my_tag`. There is no separate, dbtRunner‑specific list of arguments; the authoritative source for available options is the CLI help (`dbt --help`, `dbt run --help`, and so on) and the [dbt command reference](/reference/dbt-commands) documentation.

Reviewer: ```suggestion
For example, `dbt.invoke(["run", "--select", "tag:my_tag"])` is equivalent to running `dbt run --select tag:my_tag`. There is no separate, dbtRunner‑specific list of arguments; the authoritative source for available options is the CLI help reference (`dbt --help`, `dbt run --help`, and so on) and the [dbt command reference](/reference/dbt-commands) documentation.
```

- Running concurrent commands can unexpectedly interact with the data platform. For example, running `dbt run` and `dbt build` for the same models simultaneously could lead to unpredictable results.
- Each `dbt-core` command interacts with global Python variables. To ensure safe operation, commands need to be executed in separate processes, for example by spawning subprocesses or using Celery for orchestration.

For safe parallel execution, you can use the [<Constant name="platform_cli" />](/docs/cloud/cloud-cli-installation) or [<Constant name="studio_ide" />](/docs/cloud/studio-ide/develop-in-studio), both of which do that additional work to manage concurrency (multiple processes) on your behalf.

## `dbtRunnerResult`

Each command returns a `dbtRunnerResult` object with three attributes:

- `success` (bool): Whether the command succeeded.

> - `result`: When the command completes (successfully or with handled errors), the command's result or results. The return type varies by command.

Reviewer: ```suggestion
- `result`: When the command completes (successfully or with handled errors), it returns the command's result(s). The return type varies by command.
```
